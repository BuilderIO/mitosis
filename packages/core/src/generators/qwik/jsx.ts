import { MitosisNode } from '../../types/mitosis-node';
import { DIRECTIVES } from './directives';
import {
  File,
  INDENT,
  invoke,
  NL,
  SrcBuilder,
  quote,
  UNINDENT,
  lastProperty,
} from './src-generator';
import { CssStyles } from './styles';

export function renderJSXNodes(
  file: File,
  directives: Map<string, string>,
  handlers: Map<string, string>,
  children: MitosisNode[],
  styles: Map<string, CssStyles>,
  parentSymbolBindings: Record<string, string>,
  root = true,
): any {
  return function(this: SrcBuilder) {
    if (children.length == 0) return;
    if (root) this.emit('(', INDENT, NL);
    const needsFragment = root && children.length > 1;
    file.import(file.qwikModule, 'h');
    if (needsFragment) {
      file.import(file.qwikModule, 'Fragment');
      this.jsxBeginFragment(file.import(file.qwikModule, 'Fragment'));
    }
    children.forEach((child) => {
      if (isEmptyTextNode(child)) return;
      if (isTextNode(child)) {
        if (child.bindings._text !== undefined) {
          this.jsxTextBinding(child.bindings._text);
        } else {
          this.isJSX
            ? this.emit(child.properties._text)
            : this.jsxTextBinding(quote(child.properties._text!));
        }
      } else {
        let childName = child.name;
        const directive = DIRECTIVES[childName];
        if (typeof directive == 'function') {
          this.emit(
            directive(child, () =>
              renderJSXNodes(
                file,
                directives,
                handlers,
                child.children,
                styles,
                {},
                false,
              ).call(this),
            ),
          );
        } else {
          if (typeof directive == 'string') {
            directives.set(childName, directive);
            if (file.module !== 'med') {
              file.import('./med.js', childName);
            }
          }
          if (isSymbol(childName)) {
            // TODO(misko): We are hard coding './med.js' which is not right.
            file.import('./med.js', childName);
            let exportedChildName = file.exports.get(childName);
            if (exportedChildName) {
              childName = exportedChildName;
            }
          }
          let props: Record<string, any> = child.properties;
          const css = child.bindings.css;
          if (css) {
            props = { ...props };
            props.class = addClass(styles.get(css)!.CLASS_NAME, props.class);
          }
          const symbolBindings: Record<string, string> = {};
          const bindings = rewriteHandlers(
            file,
            handlers,
            child.bindings,
            symbolBindings,
          );
          this.jsxBegin(childName, props, {
            ...bindings,
            ...parentSymbolBindings,
          });
          renderJSXNodes(
            file,
            directives,
            handlers,
            child.children,
            styles,
            symbolBindings,
            false,
          ).call(this);
          this.jsxEnd(childName);
        }
      }
    });
    if (needsFragment) {
      this.jsxEndFragment();
    }
    if (root) this.emit(UNINDENT, ')');
  };
}

function isSymbol(name: string): boolean {
  return name.charAt(0) == name.charAt(0).toUpperCase();
}

function addClass(
  className: string,
  existingClass: string | undefined,
): string {
  return [className, ...(existingClass ? existingClass.split(' ') : [])].join(
    ' ',
  );
}

function isEmptyTextNode(child: MitosisNode) {
  return child.properties._text?.trim() == '';
}

function isTextNode(child: MitosisNode) {
  return (
    child.properties._text !== undefined || child.bindings._text !== undefined
  );
}

/**
 * Rewrites bindings:
 * - Remove `css`
 * - Rewrites event handles
 * - Extracts symbol bindings.
 *
 * @param file
 * @param handlers
 * @param bindings
 * @param symbolBindings Options record which will receive the symbol bindings
 * @returns
 */
function rewriteHandlers(
  file: File,
  handlers: Map<string, string>,
  bindings: Record<string, string | undefined>,
  symbolBindings: Record<string, string>,
): Record<string, string> {
  const outBindings: Record<string, string> = {};
  for (let key in bindings) {
    if (Object.prototype.hasOwnProperty.call(bindings, key)) {
      let binding: any = bindings[key]!;
      let handlerBlock: string | undefined;
      if (binding != null) {
        if (key == 'css') {
          continue;
        } else if ((handlerBlock = handlers.get(binding))) {
          key = `on:${key.substring(2).toLowerCase()}`;
          binding = invoke(file.import(file.qwikModule, 'qrl'), [
            quote(file.qrlPrefix + 'high.js'),
            quote(handlerBlock),
            '[__props__, __state__]',
          ]);
        } else if (symbolBindings && key.startsWith('symbol.data.')) {
          symbolBindings[lastProperty(key)] = binding;
        }
        outBindings[key] = binding;
      }
    }
  }
  return outBindings;
}
