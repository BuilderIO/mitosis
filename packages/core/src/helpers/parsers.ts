import * as babel from '@babel/core';

const jsxPlugin = require('@babel/plugin-syntax-jsx');
const tsPreset = require('@babel/preset-typescript');
const decorators = require('@babel/plugin-syntax-decorators');

export function parseCode(code: string) {
  const ast = babel.parse(code, {
    presets: [[tsPreset, { isTSX: true, allExtensions: true }]],
    plugins: [[decorators, { legacy: true }], jsxPlugin],
  });
  const body = babel.types.isFile(ast)
    ? ast.program.body
    : babel.types.isProgram(ast)
    ? ast.body
    : [];
  return body;
}

export const isCodeBodyExpression = (body: babel.types.Statement[]) =>
  body.length == 1 &&
  (babel.types.isExpression(body[0]) ||
    babel.types.isExpressionStatement(body[0]));

/**
 * Returns `true` if the `code` is a valid expression. (vs a statement)
 */
export function isExpression(code: string): boolean {
  try {
    const body = parseCode(code);
    return isCodeBodyExpression(body);
  } catch (e) {
    return false;
  }
}

export const isCodeBodyIdentifier = (body: babel.types.Statement[]) =>
  body.length == 1 && babel.types.isIdentifier(body[0]);
