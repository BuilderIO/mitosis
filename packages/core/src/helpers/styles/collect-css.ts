import traverse from 'traverse';
import { MitosisComponent } from '../../types/mitosis-component';
import { dashCase } from '../dash-case';
import { isMitosisNode } from '../is-mitosis-node';
import hash from 'object-hash';
import {
  ClassStyleMap,
  getNestedSelectors,
  getStylesOnly,
  nodeHasStyles,
  parseCssObject,
  styleMapToCss,
} from './helpers';
import { MitosisNode } from '../../types/mitosis-node';

type CollectStyleOptions = {
  prefix?: string;
};

const trimClassStr = (classStr: string) => classStr.trim().replace(/\s{2,}/g, ' ');

const updateClassForNode = (item: MitosisNode, className: string) => {
  if (item.bindings.class) {
    // combine className with existing binding. We use single quotes because in Vue, bindings are wrapped in double quotes
    // e.g. <div :class="_classStringToObject(this.className + 'div-21azgz5avex')" />
    item.bindings.class.code = trimClassStr(`${item.bindings.class.code} + '${className}'`);
  } else {
    item.properties.class = trimClassStr(`${item.properties.class || ''} ${className}`);
  }
};

const collectStyles = (
  json: MitosisComponent,
  options: CollectStyleOptions = {},
): ClassStyleMap => {
  const styleMap: ClassStyleMap = {};

  const componentIndexes: { [className: string]: number | undefined } = {};
  const componentHashes: { [className: string]: string | undefined } = {};

  traverse(json).forEach(function (item) {
    if (isMitosisNode(item)) {
      if (nodeHasStyles(item)) {
        const value = parseCssObject(item.bindings.css?.code as string);
        delete item.bindings.css;
        const componentName = item.properties.$name
          ? dashCase(item.properties.$name)
          : /^h\d$/.test(item.name || '') // don't dashcase h1 into h-1
          ? item.name
          : dashCase(item.name || 'div');

        const classNameWPrefix = `${componentName}${options.prefix ? `-${options.prefix}` : ''}`;

        const stylesHash = hash(value);
        if (componentHashes[componentName] === stylesHash) {
          const className = classNameWPrefix;
          updateClassForNode(item, className);
          return;
        }

        if (!componentHashes[componentName]) {
          componentHashes[componentName] = stylesHash;
        }

        const index = (componentIndexes[componentName] =
          (componentIndexes[componentName] || 0) + 1);
        const className = `${classNameWPrefix}${index === 1 ? '' : `-${index}`}`;

        updateClassForNode(item, className);

        styleMap[className] = value;
      }
      delete item.bindings.css;
    }
  });

  return styleMap;
};

export const collectCss = (json: MitosisComponent, options: CollectStyleOptions = {}): string => {
  const styles = collectStyles(json, options);
  // TODO create and use a root selector
  return classStyleMapToCss(styles);
};

const classStyleMapToCss = (map: ClassStyleMap): string => {
  let str = '';

  for (const key in map) {
    const styles = getStylesOnly(map[key]);
    str += `.${key} { ${styleMapToCss(styles)} }`;
    const nestedSelectors = getNestedSelectors(map[key]);
    for (const nestedSelector in nestedSelectors) {
      const value = nestedSelectors[nestedSelector] as any;
      if (nestedSelector.startsWith('@')) {
        str += `${nestedSelector} { .${key} { ${styleMapToCss(value)} } }`;
      } else {
        const useSelector = nestedSelector.includes('&')
          ? nestedSelector.replace(/&/g, `.${key}`)
          : `.${key} ${nestedSelector}`;
        str += `${useSelector} { ${styleMapToCss(value)} }`;
      }
    }
  }

  return str;
};
