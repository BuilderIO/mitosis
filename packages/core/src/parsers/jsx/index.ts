export { parseJsx } from './jsx';
export { createTypescriptProject, mapSignalTypeInTSFile } from './types-identification';
