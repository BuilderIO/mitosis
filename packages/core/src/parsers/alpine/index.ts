import { parseExpression } from '@babel/parser';
import { MitosisComponent } from '@builder.io/mitosis';
import { parse as parseHTML } from 'node-html-parser';
import { createSingleBinding } from '../../helpers/bindings';
import { createMitosisNode } from '../../helpers/create-mitosis-node';
import { MitosisNode } from '../../types/mitosis-node';
import { AlpineAST, AlpineDirective, AlpineNode } from './types';

export function parseAlpine(html: string): MitosisComponent {
  const root = parseHTML(html);
  const ast: AlpineAST = {
    type: 'AlpineAST',
    root: parseNode(root),
    data: null,
  };

  // Find x-data directive
  const dataDirective = findDataDirective(ast.root);
  if (dataDirective) {
    ast.data = {
      type: 'AlpineData',
      value: dataDirective.value,
      parsed: parseData(dataDirective.value),
    };
  }

  return convertToMitosis(ast);
}

function findDataDirective(node: AlpineNode): { value: string } | null {
  const dataDirective = node.directives.find((d) => d.name === 'x-data');
  if (dataDirective) {
    return { value: dataDirective.value };
  }

  for (const child of node.children) {
    const result = findDataDirective(child);
    if (result) {
      return result;
    }
  }

  return null;
}

function parseData(value: string): Record<string, any> {
  try {
    // First try parsing as JSON
    return JSON.parse(value);
  } catch (e) {
    try {
      // Then try parsing as JavaScript expression
      const ast = parseExpression(value, {
        sourceType: 'module',
        plugins: ['jsx', 'typescript'],
      });
      return {
        expression: ast,
        value,
      };
    } catch (e) {
      // If both fail, return the raw value
      return { value };
    }
  }
}

function parseNode(node: any): AlpineNode {
  if (node.nodeType === 3) {
    // Text node
    return {
      type: 'AlpineNode',
      tagName: '',
      attributes: {},
      directives: [],
      children: [],
      text: node.text,
    };
  }

  if (node.nodeType === 8) {
    // Comment node
    return {
      type: 'AlpineNode',
      tagName: '',
      attributes: {},
      directives: [],
      children: [],
      text: `<!--${node.text}-->`,
    };
  }

  const directives: any[] = [];
  const attributes: Record<string, string> = {};

  // Parse attributes and directives
  const attrs = node.attributes;
  if (attrs) {
    Object.keys(attrs).forEach((name) => {
      const value = attrs[name];
      if (name.startsWith('x-')) {
        const [directiveName, ...modifiers] = name.split('.');
        directives.push({
          type: 'AlpineDirective',
          name: directiveName.slice(2), // Remove 'x-' prefix
          value,
          modifiers,
        });
      } else {
        attributes[name] = value;
      }
    });
  }

  // Parse child nodes
  const children: AlpineNode[] = [];
  if (node.childNodes) {
    for (const child of node.childNodes) {
      const parsedChild = parseNode(child);
      if (parsedChild.tagName || parsedChild.text) {
        children.push(parsedChild);
      }
    }
  }

  return {
    type: 'AlpineNode',
    tagName: node.tagName?.toLowerCase() || '',
    attributes,
    directives,
    children,
    text: undefined,
  };
}

function transformDirective(directive: AlpineDirective) {
  switch (directive.name) {
    case 'text':
      return {
        _text: createSingleBinding({ code: directive.value }),
      };
    case 'bind':
      // Handle x-bind:attribute="value" or :attribute="value"
      const attribute = directive.modifiers[0];
      if (!attribute) {
        throw new Error('x-bind directive requires an attribute name');
      }
      return {
        [attribute]: createSingleBinding({ code: directive.value }),
      };
    case 'on':
      // Handle x-on:event="handler" or @event="handler"
      const event = directive.modifiers[0] || 'click';
      // Convert Alpine event modifiers to Mitosis event handlers
      const modifiers = directive.modifiers.slice(1);
      let handlerCode = directive.value;

      // Handle event modifiers
      if (modifiers.includes('prevent')) {
        handlerCode = `(e) => { e.preventDefault(); ${handlerCode} }`;
      }
      if (modifiers.includes('stop')) {
        handlerCode = `(e) => { e.stopPropagation(); ${handlerCode} }`;
      }
      if (modifiers.includes('once')) {
        handlerCode = `(function() { let called = false; return (e) => { if (!called) { called = true; ${handlerCode} } } })()`;
      }
      if (modifiers.includes('self')) {
        handlerCode = `(e) => { if (e.target === e.currentTarget) { ${handlerCode} } }`;
      }
      if (modifiers.includes('window')) {
        handlerCode = `(e) => { if (e.target === window) { ${handlerCode} } }`;
      }
      if (modifiers.includes('document')) {
        handlerCode = `(e) => { if (e.target === document) { ${handlerCode} } }`;
      }
      if (modifiers.includes('outside')) {
        handlerCode = `(e) => { if (!e.currentTarget.contains(e.target)) { ${handlerCode} } }`;
      }

      // Handle debounce and throttle
      const debounceMatch = modifiers.find((m) => m.startsWith('debounce'));
      if (debounceMatch) {
        const ms = debounceMatch.split('.')[1] || '250';
        handlerCode = `(function() { let timeout; return (e) => { clearTimeout(timeout); timeout = setTimeout(() => { ${handlerCode} }, ${ms}); } })()`;
      }

      const throttleMatch = modifiers.find((m) => m.startsWith('throttle'));
      if (throttleMatch) {
        const ms = throttleMatch.split('.')[1] || '250';
        handlerCode = `(function() { let lastCall = 0; return (e) => { const now = Date.now(); if (now - lastCall >= ${ms}) { lastCall = now; ${handlerCode} } } })()`;
      }

      return {
        [`on${event.charAt(0).toUpperCase() + event.slice(1)}`]: createSingleBinding({
          code: handlerCode,
        }),
      };
    case 'if':
      return {
        _if: createSingleBinding({ code: directive.value }),
      };
    case 'for':
      // Handle x-for="item in items" or x-for="(item, index) in items" syntax
      const forMatch =
        directive.value.match(/^\((.*?),\s*(.*?)\)\s+in\s+(.*)$/) ||
        directive.value.match(/^(.*?)\s+in\s+(.*)$/);
      if (!forMatch) {
        throw new Error('Invalid x-for syntax');
      }

      const [_, item, index, items] = forMatch;
      return {
        each: createSingleBinding({ code: items }),
        forName: item,
        indexName: index,
      };
    case 'show':
      return {
        _show: createSingleBinding({ code: directive.value }),
      };
    case 'model':
      // Handle x-model="variable" for two-way binding
      // Also handle x-model modifiers like .number, .trim, .lazy
      const modelModifiers = directive.modifiers;
      let modelValue = directive.value;
      let modelHandler = `(e) => { ${directive.value} = e.target.value }`;

      if (modelModifiers.includes('number')) {
        modelHandler = `(e) => { ${directive.value} = Number(e.target.value) }`;
      }
      if (modelModifiers.includes('trim')) {
        modelHandler = `(e) => { ${directive.value} = e.target.value.trim() }`;
      }
      if (modelModifiers.includes('lazy')) {
        modelHandler = `(e) => { ${directive.value} = e.target.value }`;
      }

      return {
        value: createSingleBinding({ code: modelValue }),
        onChange: createSingleBinding({ code: modelHandler }),
      };
    case 'html':
      return {
        innerHTML: createSingleBinding({ code: directive.value }),
      };
    case 'cloak':
      // x-cloak is handled by CSS, no need for binding
      return {};
    case 'init':
      // x-init is handled by component initialization
      return {};
    case 'transition':
      // Handle x-transition directives
      const transitionModifier = directive.modifiers[0];
      if (transitionModifier) {
        return {
          [`transition${transitionModifier.charAt(0).toUpperCase() + transitionModifier.slice(1)}`]:
            createSingleBinding({ code: directive.value }),
        };
      }
      return {};
    case 'teleport':
      // Handle x-teleport directive
      return {
        _teleport: createSingleBinding({ code: directive.value }),
      };
    case 'ref':
      // Handle x-ref directive
      return {
        ref: createSingleBinding({ code: directive.value }),
      };
    default:
      return {
        [directive.name]: createSingleBinding({ code: directive.value }),
      };
  }
}

function transformNode(node: AlpineNode | null | undefined): MitosisNode {
  // Handle null or undefined nodes
  if (!node) {
    return createMitosisNode({
      '@type': '@builder.io/mitosis/node',
      name: 'div',
      properties: {},
      bindings: {},
      children: [],
      meta: {},
      scope: {},
    });
  }

  // Handle text and comment nodes
  if (node.text) {
    return createMitosisNode({
      '@type': '@builder.io/mitosis/node',
      name: 'div',
      properties: {},
      bindings: {
        _text: createSingleBinding({ code: `"${node.text}"` }),
      },
      children: [],
      meta: {},
      scope: {},
    });
  }

  const mitosisNode = createMitosisNode({
    '@type': '@builder.io/mitosis/node',
    name: node.tagName || 'div',
    properties: node.attributes || {},
    bindings: {},
    children: [],
    meta: {},
    scope: {},
  });

  // Transform directives
  if (node.directives) {
    node.directives.forEach((directive) => {
      Object.assign(mitosisNode.bindings, transformDirective(directive));
    });
  }

  // Transform children
  if (node.children) {
    mitosisNode.children = node.children.map((child) => transformNode(child));
  }

  return mitosisNode;
}

function convertToMitosis(ast: AlpineAST): MitosisComponent {
  const component: MitosisComponent = {
    '@type': '@builder.io/mitosis/component',
    name: 'AlpineComponent',
    imports: [],
    state: ast.data?.parsed || {},
    children: [],
    meta: {},
    hooks: {
      onMount: [],
      onEvent: [],
    },
    subComponents: [],
    context: {
      get: {},
      set: {},
    },
    props: {},
    refs: {},
    inputs: [],
  };

  component.children = [transformNode(ast.root)];

  return component;
}
