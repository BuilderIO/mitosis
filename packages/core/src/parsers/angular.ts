import ts from 'typescript';
import { parseTemplate } from '@angular/compiler';
import {
  Node,
  Element,
  Template,
  Text,
  BoundText,
} from '@angular/compiler/src/render3/r3_ast';
import { ASTWithSource } from '@angular/compiler/src/expression_parser/ast';
import { createMitosisComponent } from '../helpers/create-mitosis-component';
import { createMitosisNode } from '../helpers/create-mitosis-node';
import { MitosisNode } from '../types/mitosis-node';
import { omit } from 'lodash';
import { babelTransformCode } from '../helpers/babel-transform';
import { types } from '@babel/core';
import { MitosisComponent } from '../types/mitosis-component';
import { capitalize } from '../helpers/capitalize';

const getTsAST = (code: string) => {
  return ts.createSourceFile('code.ts', code, ts.ScriptTarget.Latest, true);
};

interface AngularToMitosisOptions {}

const transformBinding = (
  binding: string,
  _options: AngularToMitosisOptions,
) => {
  return babelTransformCode(binding, {
    Identifier(path) {
      const name = path.node.name;

      if (
        (types.isObjectProperty(path.parent) &&
          path.parent.key === path.node) ||
        (types.isMemberExpression(path.parent) &&
          path.parent.property === path.node)
      ) {
        return;
      }

      if (
        !(name.startsWith('state.') || name === 'event' || name === '$event')
      ) {
        path.replaceWith(types.identifier(`state.${name}`));
      }
    },
  });
};

const isElement = (node: Node): node is Element =>
  // TODO: theres got to be a better way than this
  Array.isArray((node as any).attributes);

const isTemplate = (node: Node): node is Template =>
  // TODO: theres got to be a better way than this
  Array.isArray((node as any).templateAttrs);

const isText = (node: Node): node is Text =>
  typeof (node as any).value === 'string';

const isBoundText = (node: Node): node is BoundText =>
  typeof (node as any).value === 'object';

const angularTemplateNodeToMitosisNode = (
  node: Node,
  options: AngularToMitosisOptions,
): MitosisNode => {
  if (isTemplate(node)) {
    const ngIf = node.templateAttrs.find((item) => item.name === 'ngIf');
    if (ngIf) {
      return createMitosisNode({
        name: 'Show',
        bindings: {
          when: {
            code: transformBinding(
              (ngIf.value as ASTWithSource).source!,
              options,
            ),
          },
        },
        children: [
          angularTemplateNodeToMitosisNode(
            omit(node, 'templateAttrs'),
            options,
          ),
        ],
      });
    }
    const ngFor = node.templateAttrs.find((item) => item.name === 'ngFor');
    if (ngFor) {
      const value = (ngFor.value as ASTWithSource).source!;
      const split = value.split(/let\s|\sof\s/);
      const [_let, itemName, _of, expression] = split;
      return createMitosisNode({
        name: 'For',
        bindings: {
          each: { code: transformBinding(expression, options) },
        },
        properties: {
          _forName: itemName,
        },
        children: [
          angularTemplateNodeToMitosisNode(
            omit(node, 'templateAttrs'),
            options,
          ),
        ],
      });
    }
  }

  if (isElement(node)) {
    const properties: Record<string, string> = {};
    const bindings: Record<string, { code: string; arguments?: string[] }> = {};

    for (const input of node.inputs) {
      bindings[input.name] = {
        code: transformBinding((input.value as ASTWithSource).source!, options),
      };
    }
    for (const output of node.outputs) {
      bindings['on' + capitalize(output.name)] = {
        code: transformBinding(
          (output.handler as ASTWithSource)
            .source! // TODO: proper reference replace
            .replace(/\$event/g, 'event'),
          options,
        ),
      };
    }
    for (const attribute of node.attributes) {
      properties[attribute.name] = attribute.value;
    }

    return createMitosisNode({
      name: node.name,
      properties,
      bindings,
      children: node.children.map((node) =>
        angularTemplateNodeToMitosisNode(node, options),
      ),
    });
  }

  if (isText(node)) {
    return createMitosisNode({
      properties: {
        _text: node.value,
      },
    });
  }

  if (isBoundText(node)) {
    // TODO: handle the bindings
    return createMitosisNode({
      properties: {
        _text: (node.value as ASTWithSource).source!,
      },
    });
  }

  throw new Error(`Element node type {${node}} is not supported`);
};

const angularTemplateToMitosisNodes = (
  template: string,
  options: AngularToMitosisOptions,
) => {
  const ast = parseTemplate(template, '.');
  const blocks = ast.nodes.map((node) =>
    angularTemplateNodeToMitosisNode(node, options),
  );

  return blocks;
};

const parseTypescript = (code: string, options: AngularToMitosisOptions) => {
  const component = createMitosisComponent();

  const ast = getTsAST(code);
  for (const statement of ast.statements) {
    if (ts.isClassDeclaration(statement)) {
      if (statement.decorators) {
        for (const decorator of statement.decorators) {
          // TODO: proper reference tracing
          if (ts.isCallExpression(decorator.expression))
            if (
              ts.isIdentifier(decorator.expression.expression) &&
              decorator.expression.expression.text === 'Component'
            ) {
              const firstArg = decorator.expression.arguments[0];
              if (ts.isObjectLiteralExpression(firstArg)) {
                firstArg.properties.find((item) => {
                  if (ts.isPropertyAssignment(item)) {
                    if (
                      ts.isIdentifier(item.name) &&
                      item.name.text === 'template'
                    ) {
                      if (ts.isTemplateLiteral(item.initializer)) {
                        const template = item.initializer
                          .getText()
                          .trim()
                          .slice(1, -1);
                        component.children = angularTemplateToMitosisNodes(
                          template,
                          options,
                        );
                      }
                    }
                  }
                });
              }
            }
        }
      }
    }
  }

  return component;
};

export function angularToMitosisComponent(
  code: string,
  options: AngularToMitosisOptions = {},
): MitosisComponent {
  return parseTypescript(code, options);
}
