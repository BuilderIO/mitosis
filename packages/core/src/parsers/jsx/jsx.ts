import * as babel from '@babel/core';
import generate from '@babel/generator';
import { pipe } from 'fp-ts/lib/function';
import { createMitosisComponent } from '../../helpers/create-mitosis-component';
import { tryParseJson } from '../../helpers/json';
import { stripNewlinesInStrings } from '../../helpers/replace-new-lines-in-strings';
import { MitosisComponent } from '../../types/mitosis-component';
import { jsonToAst } from './ast';
import { collectTypes, isTypeOrInterface } from './component-types';
import { extractContextComponents } from './context';
import { jsxElementToJson } from './element-parser';
import { generateExports } from './exports';
import { componentFunctionToJson } from './function-parser';
import { isImportOrDefaultExport } from './helpers';
import { collectModuleScopeHooks } from './hooks';
import { handleImportDeclaration } from './imports';
import { undoPropsDestructure } from './props';
import { mapStateIdentifiers } from './state';
import { Context, ParseMitosisOptions } from './types';

import tsPlugin from '@babel/plugin-syntax-typescript';
import tsPreset from '@babel/preset-typescript';
import { HOOKS } from '../../constants/hooks';
import { getMagicString, getTargetId, getUseTargetStatements } from './hooks/use-target';

const { types } = babel;

const typescriptBabelPreset = [tsPreset, { isTSX: true, allExtensions: true }];

const beforeParse = (path: babel.NodePath<babel.types.Program>) => {
  path.traverse({
    FunctionDeclaration(path) {
      undoPropsDestructure(path);
    },
  });
};

/**
 * This function takes the raw string from a Mitosis component, and converts it into a JSON that can be processed by
 * each generator function.
 *
 * @param jsx string representation of the Mitosis component
 * @returns A JSON representation of the Mitosis component
 */
export function parseJsx(
  jsx: string,
  _options: Partial<ParseMitosisOptions> = {},
): MitosisComponent {
  let subComponentFunctions: string[] = [];

  const options: ParseMitosisOptions = {
    typescript: false,
    ..._options,
  };

  const jsxToUse = options.typescript
    ? jsx
    : // strip typescript types by running through babel's TS preset.
      (babel.transform(jsx, {
        configFile: false,
        babelrc: false,
        presets: [typescriptBabelPreset],
      })?.code as string);

  const output = babel.transform(jsxToUse, {
    configFile: false,
    babelrc: false,
    comments: false,
    plugins: [
      [tsPlugin, { isTSX: true }],
      (): babel.PluginObj<Context> => ({
        visitor: {
          JSXExpressionContainer(path, context) {
            if (types.isJSXEmptyExpression(path.node.expression)) {
              path.remove();
            }
          },
          Program(path, context) {
            if (context.builder) {
              return;
            }

            beforeParse(path);

            context.builder = {
              component: createMitosisComponent(),
            };

            const keepStatements = path.node.body.filter(
              (statement) => isImportOrDefaultExport(statement) || isTypeOrInterface(statement),
            );

            context.builder.component.exports = generateExports(path);

            subComponentFunctions = path.node.body
              .filter(
                (node) =>
                  !types.isExportDefaultDeclaration(node) && types.isFunctionDeclaration(node),
              )
              .map((node) => `export default ${generate(node).code!}`);

            const preComponentCode = pipe(
              path.node.body.filter((statement) => !isImportOrDefaultExport(statement)),
              collectModuleScopeHooks(context.builder.component, options),
              types.program,
              generate,
              (generatorResult) => generatorResult.code,
            );

            // TODO: support multiple? e.g. for others to add imports?
            context.builder.component.hooks.preComponent = { code: preComponentCode };

            path.replaceWith(types.program(keepStatements));
          },
          FunctionDeclaration(path, context) {
            const { node } = path;
            if (types.isIdentifier(node.id)) {
              const name = node.id.name;
              if (name[0].toUpperCase() === name[0]) {
                path.traverse({
                  /**
                   * Plugin to find all `useTarget()` assignment calls inside of the component function body
                   * and replace them with a magic string.
                   */
                  CallExpression(path) {
                    if (!types.isVariableDeclarator(path.parent)) return;
                    if (!types.isCallExpression(path.node)) return;
                    if (!types.isIdentifier(path.node.callee)) return;
                    if (path.node.callee.name !== HOOKS.TARGET) return;

                    const targetBlock = getUseTargetStatements(path.node);

                    if (!targetBlock) return;

                    const blockId = getTargetId(context.builder.component);

                    // replace the useTarget() call with a magic string
                    path.replaceWith(types.stringLiteral(getMagicString(blockId)));

                    // store the target block in the component
                    context.builder.component.targetBlocks = {
                      ...context.builder.component.targetBlocks,
                      [blockId]: targetBlock,
                    };
                  },
                });
                path.replaceWith(jsonToAst(componentFunctionToJson(node, context)));
              }
            }
          },
          ImportDeclaration(path, context) {
            handleImportDeclaration({ options, path, context });
          },
          ExportDefaultDeclaration(path) {
            path.replaceWith(path.node.declaration);
          },
          JSXElement(path) {
            const { node } = path;
            path.replaceWith(jsonToAst(jsxElementToJson(node)));
          },
          ExportNamedDeclaration(path, context) {
            const { node } = path;
            if (
              babel.types.isTSInterfaceDeclaration(node.declaration) ||
              babel.types.isTSTypeAliasDeclaration(node.declaration)
            ) {
              collectTypes(path, context);
            }
          },
          TSTypeAliasDeclaration(path, context) {
            collectTypes(path, context);
          },
          TSInterfaceDeclaration(path, context) {
            collectTypes(path, context);
          },
        },
      }),
    ],
  });

  const stringifiedMitosisComponent = stripNewlinesInStrings(
    output!
      .code!.trim()
      // Occasional issues where comments get kicked to the top. Full fix should strip these sooner
      .replace(/^\/\*[\s\S]*?\*\/\s*/, '')
      // Weird bug with adding a newline in a normal at end of a normal string that can't have one
      // If not one-off find full solve and cause
      .replace(/\n"/g, '"')
      .replace(/^\({/, '{')
      .replace(/}\);$/, '}'),
  );

  const mitosisComponent = tryParseJson(stringifiedMitosisComponent);

  mapStateIdentifiers(mitosisComponent);
  extractContextComponents(mitosisComponent);

  mitosisComponent.subComponents = subComponentFunctions.map((item) => parseJsx(item, options));

  return mitosisComponent;
}
