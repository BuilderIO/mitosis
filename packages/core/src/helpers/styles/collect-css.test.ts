import { parseJsx } from '../../parsers/jsx';
import { collectCss } from './collect-css';

import classRaw from '../../__tests__/data/styles/class.raw.tsx?raw';
import classState from '../../__tests__/data/styles/classState.raw.tsx?raw';

describe('Styles', () => {
  test('class property and CSS are merged', () => {
    const component = parseJsx(classRaw);
    const output = collectCss(component);
    expect({ component, output }).toMatchSnapshot();
  });
  test('class binding and CSS are merged', () => {
    const component = parseJsx(classState);
    const output = collectCss(component);
    expect({ component, output }).toMatchSnapshot();
  });
});
