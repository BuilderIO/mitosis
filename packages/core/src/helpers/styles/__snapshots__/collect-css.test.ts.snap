// Vitest Snapshot v1

exports[`Styles > class binding and CSS are merged 1`] = `
{
  "component": {
    "@type": "@builder.io/mitosis/component",
    "children": [
      {
        "@type": "@builder.io/mitosis/node",
        "bindings": {
          "class": {
            "code": "state.classState + ' div'",
            "type": "single",
          },
          "style": {
            "code": "state.styleState",
            "type": "single",
          },
        },
        "children": [
          {
            "@type": "@builder.io/mitosis/node",
            "bindings": {},
            "children": [],
            "meta": {},
            "name": "div",
            "properties": {
              "_text": "
      Hello! I can run in React, Vue, Solid, or Liquid!
    ",
            },
            "scope": {},
          },
        ],
        "meta": {},
        "name": "div",
        "properties": {},
        "scope": {},
      },
    ],
    "context": {
      "get": {},
      "set": {},
    },
    "exports": {},
    "hooks": {},
    "imports": [],
    "inputs": [],
    "meta": {},
    "name": "MyBasicComponent",
    "refs": {},
    "state": {
      "classState": {
        "code": "'testClassName'",
        "propertyType": "normal",
        "type": "property",
      },
      "styleState": {
        "code": "{
  color: 'red'
}",
        "propertyType": "normal",
        "type": "property",
      },
    },
    "subComponents": [],
  },
  "output": ".div {
  padding: 10px;
}",
}
`;

exports[`Styles > class property and CSS are merged 1`] = `
{
  "component": {
    "@type": "@builder.io/mitosis/component",
    "children": [
      {
        "@type": "@builder.io/mitosis/node",
        "bindings": {},
        "children": [
          {
            "@type": "@builder.io/mitosis/node",
            "bindings": {},
            "children": [],
            "meta": {},
            "name": "div",
            "properties": {
              "_text": "
      Hello! I can run in React, Vue, Solid, or Liquid!
    ",
            },
            "scope": {},
          },
        ],
        "meta": {},
        "name": "div",
        "properties": {
          "class": "test div",
        },
        "scope": {},
      },
    ],
    "context": {
      "get": {},
      "set": {},
    },
    "exports": {},
    "hooks": {},
    "imports": [],
    "inputs": [],
    "meta": {},
    "name": "MyBasicComponent",
    "refs": {},
    "state": {},
    "subComponents": [],
  },
  "output": ".div {
  padding: 10px;
}",
}
`;
